### PR TITLE
perf-tuning: add document for predicate push down

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What is changed, added or deleted? (Required)
Add document for logical plan optimization rule: predication push down.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [x] v3.1 (TiDB 3.1 versions)
- [x] v3.0 (TiDB 3.0 versions)
- [x] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### CODE REVIEW

File: predicate-push-down.md [hunk 1]
```diff
@@ -3,4 +3,142 @@ title: 谓词下推
 category: performance
 ---
 
-# 谓词下推
\ No newline at end of file
+# 谓词下推
+
```
Comment (ran-huang) [inline predicate-push-down.md:7]: Please add a sentence to describe the content of this document. For example:
“本文档介绍……”

File: predicate-push-down.md [hunk 2]
```diff
@@ -3,4 +3,142 @@ title: 谓词下推
...
+谓词下推将查询语句中的过滤表达式计算尽可能下推到距离数据源最近的地方，以尽早完成数据的过滤，进而显著地减少数据传输或计算的开销。
+
+## 示例
+
+以下通过一些例子对谓词下推优化进行说明，其中示例1/2/3为谓词下推适用的案例，示例4/5/6为谓词下推不适用的案例。
+
+### 示例 1
```
Comment (ran-huang) [inline predicate-push-down.md:14]: Can we use a more specific heading than this? Like, specify which type of predicates.

File: predicate-push-down.md [hunk 3]
```diff
@@ -3,4 +3,142 @@ title: 谓词下推
...
+|   ├─TableReader_13(Build)     | 10000.00 | root      |               | data:TableFullScan_12                           |
+|   │ └─TableFullScan_12        | 10000.00 | cop[tikv] | table:s       | keep order:false, stats:pseudo                  |
+|   └─TableReader_11(Probe)     | 10000.00 | root      |               | data:TableFullScan_10                           |
+|     └─TableFullScan_10        | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo                  |
++-------------------------------+----------+-----------+---------------+-------------------------------------------------+
+6 rows in set (0.00 sec)
+```
+
+在该查询中，内表 s 上存在谓词 `s.a is null`。
+从 explain 中可以看到，该谓词没有被下推到 join 前进行计算，这是因为外连接在不满足 on 条件时会对内表填充 NULL，而在该查询中 `s.a is null` 用来对 join 后的结果进行过滤，如果将其下推到 join 前在内表上进行过滤，则下推前后不等价, 因此不可进行下推。
```
Comment (ran-huang) [inline predicate-push-down.md:107]:
```suggestion
在该查询中，内表 s 上存在谓词 `s.a is null`。

从 explain 中可以看到，该谓词没有被下推到 join 前进行计算，这是因为外连接在不满足 on 条件时会对内表填充 NULL，而在该查询中 `s.a is null` 用来对 join 后的结果进行过滤，如果将其下推到 join 前在内表上进行过滤，则下推前后不等价，因此不可进行下推。
```

File: predicate-push-down.md [hunk 4]
```diff
@@ -3,4 +3,142 @@ title: 谓词下推
...
 
-# 谓词下推
\ No newline at end of file
+# 谓词下推
+
+谓词下推将查询语句中的过滤表达式计算尽可能下推到距离数据源最近的地方，以尽早完成数据的过滤，进而显著地减少数据传输或计算的开销。
+
+## 示例
+
+以下通过一些例子对谓词下推优化进行说明，其中示例1/2/3为谓词下推适用的案例，示例4/5/6为谓词下推不适用的案例。
```
Comment (zz-jason) [inline predicate-push-down.md:12]:
```suggestion
以下通过一些例子对谓词下推优化进行说明，其中示例 1、2、3 为谓词下推适用的案例，示例 4、5、6 为谓词下推不适用的案例。
```

File: predicate-push-down.md [hunk 5]
```diff
@@ -3,4 +3,142 @@ title: 谓词下推
...
++-------------------------+----------+-----------+---------------+--------------------------------+
+| id                      | estRows  | task      | access object | operator info                  |
++-------------------------+----------+-----------+---------------+--------------------------------+
+| TableReader_7           | 3323.33  | root      |               | data:Selection_6               |
+| └─Selection_6           | 3323.33  | cop[tikv] |               | lt(test.t.a, 1)                |
+|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo |
++-------------------------+----------+-----------+---------------+--------------------------------+
+```
+
+该查询与示例 1 中的查询生成了完成一样的执行计划，这是因为谓词 `a < substring('123', 1, 1)` 的 `substring` 的入参均为常量，因此可以提前计算，进而简化得到等价的谓词 `a < 1`。进一步的，可以将 `a < 1` 下推至 TiKV 上。
```
Comment (zz-jason) [inline predicate-push-down.md:45]: 是否新开一个章节单独介绍下常量折叠？

File: predicate-push-down.md [hunk 6]
```diff
@@ -3,4 +3,142 @@ title: 谓词下推
...
+| │ └─Selection_14             | 3323.33  | cop[tikv] |               | lt(test.s.a, 1)                            |
+| │   └─TableFullScan_13       | 10000.00 | cop[tikv] | table:s       | keep order:false, stats:pseudo             |
+| └─TableReader_12(Probe)      | 3323.33  | root      |               | data:Selection_11                          |
+|   └─Selection_11             | 3323.33  | cop[tikv] |               | lt(test.t.a, 1)                            |
+|     └─TableFullScan_10       | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo             |
++------------------------------+----------+-----------+---------------+--------------------------------------------+
+7 rows in set (0.00 sec)
+```
+
+在该查询中，将谓词 `t1.a < 1` 下推到 join 前进行过滤，可以减少 join 时的计算开销。
```
Comment (zz-jason) [inline predicate-push-down.md:67]:
```suggestion
在该查询中，将谓词 `t.a < 1` 下推到 join 前对表 `t` 进行过滤，可以减少 join 时的计算开销。
```

File: predicate-push-down.md [hunk 7]
```diff
@@ -3,4 +3,142 @@ title: 谓词下推
...
+此外，通过 `t.a = s.a`, 构造出谓词 `s.a < 1`，并将其下推至 join 运算前，可以进一步减少 join 时的计算开销。
```
Comment (zz-jason) [inline predicate-push-down.md:68]:
```suggestion
此外，这条 SQL 执行的是内连接，且 `ON` 条件时 `t.a = s.a`，可以由 `t.a < 1` 推导出谓词 `s.a < 1`，并将其下推至 join 运算前对 `s` 表进行过滤，可以进一步减少 join 时的计算开销。
```

**Comment by ran-huang** [comment]
@XuHuaiyu Please address comments. Thanks!

File: predicate-push-down.md (hunk 5 above)
Comment (XuHuaiyu) [reply predicate-push-down.md:45]: 感觉实际上这句话能解释的清？
`这是因为谓词 `a < substring('123', 1, 1)` 的 `substring` 的入参均为常量，因此可以提前计算`

**Comment by ran-huang** [comment]
@zz-jason PTAL

**Pull request merged**